Mas i1804 peerdiscovery

martinsumner opened the pull request:
See https://github.com/basho/riak_kv/issues/1804

The heart of the problem is how to avoid needing configuration changes on sink clusters when source clusters are bing changed.  This allows for new nodes to be discovered automatically, from configured nodes.

Default behaviour is to always fallback to configured behaviour.

Worker Counts and Per Peer Limits need to be set based on an understanding of whether this will be enabled.  Although, if per peer limit is left to default, the consequence will be the worker count will be evenly distributed (independently by each node).  Note, if Worker Count mod (Src Node Count) =/= 0 - then there will be no balancing of the excess workers across the sink nodes.

martinsumner committed:
Initial peer discovery code

Refactoring of riak_kv_replrtq to allow sharing of code and interaction between snk and peer

martinsumner committed:
Add membership_request (pb)

martinsumner committed:
Update riak_client.erl

martinsumner committed:
Update following test

Adds some further logging.  Also corrects the comparison between current and discovered peers to avoid unnecessary resets.

martinsumner committed:
Updates following extension of test

Adds operator riak_client functions

martinsumner committed:
Add HTTP API for membership_request

martinsumner committed:
Close clients in work queue when terminating

martinsumner committed:
Update riak_kv_replrtq_snk.erl

martinsumner committed:
Add configuration items

martinsumner committed:
Update riak_kv_replrtq_peer.erl

ThomasArts commented on src/riak_kv_replrtq_peer.erl:
```diff
@@ -0,0 +1,240 @@
+%% Prompt for the discovery of peers
+-spec update_discovery(riak_kv_replrtq_snk:queue_name()) -> boolean().
+update_discovery(QueueName) ->
+    gen_server:call(?MODULE, {update_discovery, QueueName}, 60 * 1000).
```
Do you really want 60 hardcoded here or rather use the macro `?DISCOVERY_TIMEOUT_SECONDS`.

ThomasArts commented on src/riak_kv_replrtq_peer.erl:
```diff
@@ -0,0 +1,240 @@
+                fun({QueueName, _PeerInfo}) -> 
+                    erlang:send_after(MinDelay * 1000,
+                        self(),
+                        {prompt_discovery, QueueName})
```
It takes a few moments to realize that there are 2 interfaces for `prompt discovery` where this one goes via `handle_info` and has only one argument that does a lookup for the PeerInfo at the time it executes, whereas the cast version gets the PeerInfo in the interface.


Probably one should comment this difference more clearly. In particular why one does not need to do a state lookup for the cast version

ThomasArts commented on src/riak_kv_replrtq_peer.erl:
```diff
@@ -0,0 +1,240 @@
+handle_info({prompt_discovery, QueueName}, State) ->
+    {QueueName, PeerInfo} =
+        lists:keyfind(QueueName, 1, State#state.discovery_peers),
+    ok = prompt_discovery({QueueName, PeerInfo}),
```
I kind of dislike this... you leave the context of the genserver in the cast to later return to it... I wonder if it would not be better to actually call `prompt_discovery(QueueName, PeerInfo, regular)` at this point.

ThomasArts commented on src/riak_kv_replrtq_peer.erl:
```diff
@@ -0,0 +1,240 @@
+                        riak_kv_replrtq_snk:current_peers(QueueName)))
+        end,
+    case discover_peers(PeerInfo, StartDelayMS) of
+        CurrentPeers ->
```
I guess this is by design, but if you run with Type `count_change` then CurrentPeers is the empty list. So in case discover_peers returns an empty list, you don't know in which of the two cases you are.

Both cases return false, but the side effects are different.

martinsumner replied:
Tried to avoid this confusion by not relying on the empty list mis-match - rather a specific mis-match between list and atom.  Commented as well.  Don't think it is super-clean still, but improved maybe.

ThomasArts commented on src/riak_client.erl:
```diff
@@ -137,6 +138,51 @@ maybe_update_consistent_stat(Node, Stat, Bucket, StartTS, Result) ->
+    non_neg_integer()) -> list(node()).
+replrtq_reset_all_workercounts(WorkerC, PerPeerL) ->
+    UpNodes = riak_core_node_watcher:nodes(riak_kv),
+    lists:foldl(replrtq_resetcount_fun(WorkerC, PerPeerL), [], UpNodes).
```
Personally I find the code clearer if you inline this function.

ThomasArts commented on src/riak_kv_replrtq_snk.erl:
```diff
@@ -409,13 +428,45 @@ handle_info({prompt_requeue, WorkItem}, State) ->
     {noreply, State}.
 
-terminate(_Reason, _State) ->
+terminate(_Reason, State) ->
```
This terminate function may now take considerably more time.

If the supervisor terminates this server, how likely is it that this takes too long compared to the time you want it restarted?

ThomasArts commented on src/riak_kv_replrtq_snk.erl:
```diff
@@ -409,13 +428,45 @@ handle_info({prompt_requeue, WorkItem}, State) ->
+    {SnkWorkerCount, PerPeerLimit}.
+
+set_worker_counts(SnkWorkerCount, PerPeerLimit) ->
+    application:set_env(riak_kv, replrtq_sinkworkers, SnkWorkerCount),
```
Why using app_helper for getting and application for setting? Is this an artefcat of OTP18 or so?

martinsumner replied:
Never thought about this much, just following inline with its use elsewhere.  Looking at the docs application:get_env/3 didn't exist until R16B ... and Riak was initially written prior to that ... so I guess it is just a legacy of this.

It might be one thing to add to the list for the OTP 24+ version of Riak, to refactor throughout and to take this out.

ThomasArts reviewed (approved):
Some smaller comments, nothing structural, just making sure you did things on purpose.

martinsumner committed:
Updates following code review

Rename _discovery functions to reduce confusion over multiple things called prompt_discovery doing different things

martinsumner committed:
Update to avoid https/git issue

martinsumner committed:
Convert app_helper -> application in replrtq

martinsumner committed:
app_helper:get/2 =/= application:get_env/2

Saves some pain with pattern matching {ok, V}|undefined

martinsumner committed:
Revert back as app_helper:get_env/2 =/= appplication:get_env/2

martinsumner committed:
Update riak_kv_replrtq_peer.erl
